### PR TITLE
Environment Optimization

### PR DESCRIPTION
**Pull request by MarkHaoxiang**
Some low hanging fruit to optimize the environment. With low population counts, up to 50% faster.

### CODE REVIEW

**Comment by kywch** [comment]
Thanks, Mark! There seems to be many changes here and there, so allow me to take a careful look.

File: nmmo/core/map.py [hunk 1]
```diff
@@ -62,6 +64,7 @@ def reset(self, map_id, np_random):
         mat  = materials[idx]
         tile = self.tiles[r, c]
         tile.reset(mat, config, np_random)
+        self.habitable_tiles[r, c] = tile.habitable
```
Comment (kywch) [inline nmmo/core/map.py:67]: nice!

File: nmmo/systems/ai/move.py [hunk 2]
```diff
@@ -2,12 +2,13 @@
 from nmmo.core import action
 from nmmo.systems.ai import utils
 
-
+# pylint: disable=E1136
```
Comment (kywch) [inline nmmo/systems/ai/move.py]: please change to the error category (the same for the first line)

File: nmmo/systems/ai/move.py [hunk 3]
```diff
@@ -20,22 +21,48 @@ def random_safe(tiles, ent, np_random):
 
   return np_random.choice(cands)
 
-def habitable(tiles, ent, np_random):
+def habitable(realm_map, ent, np_random):
```
Comment (kywch) [inline nmmo/systems/ai/move.py:11]: how much speed do we gain by changing `np_random.choice(cands)` to below? 

also, would it change a lot if we replace habitable with random_safe?

File: nmmo/systems/ai/utils.py [hunk 4]
```diff
@@ -50,10 +47,6 @@ def closestTarget(ent, tiles, rng=1):
          for e in tiles[sr + d, sc + r].entities.values():
             if e is not ent and validTarget(ent, e, rng): return e
 
-def distance(ent, targ):
-   # used in scripted/behavior.py, attack() to determine attack range
-   return lInfty(ent.pos, targ.pos)
-
 def lInf(ent, targ):
```
Comment (kywch) [inline nmmo/systems/ai/utils.py:56]: can we also get rid of this function?

File: nmmo/systems/ai/utils.py [hunk 5]
```diff
@@ -8,18 +8,15 @@
...
-   if lInfty(ent.pos, targ.pos) > rng:
+   if targ is None or not targ.alive or lInfty(ent.pos, targ.pos) > rng:
       return False
    return True
 
 
 def validResource(ent, tile, rng):
    return tile is not None and tile.state.tex in (
-      'foilage', 'water') and distance(ent, tile) <= rng
+      'foilage', 'water') and lInfty(ent.pos, tile.pos) <= rng
```
Comment (kywch) [inline nmmo/systems/ai/utils.py]: can we use `utils.linf_single` here? 

there are several distance functions defined, so it'd be great if we can remove redundant ones

File: nmmo/systems/ai/utils.py [hunk 6]
```diff
@@ -88,10 +81,15 @@ def lInfty(start, goal):
...
 
-def aStar(tiles, start, goal, cutoff=100):
+CUTOFF = 100
+def aStar(realm_map, start, goal):
+   cutoff = CUTOFF
+   tiles = realm_map.tiles
    if start == goal:
       return (0, 0)
-
+   if (start,goal) in realm_map.pathfinding_cache:
```
Comment (kywch) [inline nmmo/systems/ai/utils.py]: nice!

File: tests/test_optimization.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
+# Test invariants assumed for certain optimizations
```
Comment (kywch) [inline tests/test_optimization.py]: please move this test file to `tests/core/test_immutable_tile_property.py` and rename accordingly

**Review by kywch — CHANGES_REQUESTED** [review]
Thank you for making npcs much faster, Mark! The test I've been using confirms the speed up.

```python
from nmmo.task.task_api import nmmo_default_task
from tests.testhelpers import profile_env_step

# nmmo.config.Default() setting: 128 agents
tasks = nmmo_default_task(range(128))
profile_env_step(tasks=tasks)
```

The 2.0 branch:
```
 - env.step({}): 5.237950573999115
 - env.realm.step(): 1.8601984710003308
 - env._compute_observations(): 0.6015106539998669
 - obs.to_gym(), ActionTarget: 0.958669548999751
 - env._compute_rewards(): 0.8221662120013207
```

The current branch:
```
 - env.step({}): 4.958883338000305
 - env.realm.step(): 1.3220626919974166   <-- YAY
 - env._compute_observations(): 0.5937297150012455
 - obs.to_gym(), ActionTarget: 0.9411229560028005
 - env._compute_rewards(): 0.9482983250018151
```

However, removing the game state indices, caches seem to hurt the performance, when tested with 128 players and 128 npcs, the numbers we use in the competition. Can you please confirm if the game state-related changes are necessary?

Thank you.

**Comment by kywch** [comment]
Here is another test case to optimize, simulating complex tasks, touching all entity, item, and event data. 

```python
  from nmmo.task.task_api import nmmo_default_task, make_same_task
  from nmmo.task.base_predicates import CountEvent, FullyArmed
  from nmmo.systems.skill import Melee
  from tests.testhelpers import profile_env_step

  # nmmo.config.Default() setting: 128 agents
  tasks = nmmo_default_task(range(128))
  tasks += make_same_task(CountEvent, range(128),
                          pred_kwargs={'event': 'EAT_FOOD', 'N': 10})
  tasks += make_same_task(FullyArmed, range(128),
                          pred_kwargs={'combat_style': Melee, 'level': 3, 'num_agent': 1})
  profile_env_step(tasks=tasks)
```

2.0 branch
```
 - env.step({}): 9.356640450998384
 - env.realm.step(): 1.9452383820025716
 - env._compute_observations(): 0.6063221270014765
 - obs.to_gym(), ActionTarget: 0.9888858630001778
 - env._compute_rewards(): 4.478575074994296
```

this branch
```
 - env.step({}): 9.688806402999035
 - env.realm.step(): 1.4039403799979482
 - env._compute_observations(): 0.606985855003586
 - obs.to_gym(), ActionTarget: 0.998996601003455
 - env._compute_rewards(): 5.417658225997002
```

As the (training) task becomes complex, computing rewards takes so much time. This certainly is a big, separate target for optimization, after merging in the nice gains in this PR.

**Comment by kywch** [comment]
With a custom function that combines StayAlive, CountEvent, FullyArmed

```python
  from nmmo.task.task_api import nmmo_default_task, make_same_task
  from nmmo.task.base_predicates import StayAlive, CountEvent, FullyArmed
  from nmmo.systems.skill import Melee
  from tests.testhelpers import profile_env_step

  def complex_task(gs, subject, event, N, combat_style, level, num_agent):
    return StayAlive(gs, subject) * CountEvent(gs, subject, event, N) * \
           FullyArmed(gs, subject, combat_style, level, num_agent)

  # nmmo.config.Default() setting: 128 agents
  tasks = make_same_task(complex_task, range(128),
                         pred_kwargs={'event': 'EAT_FOOD', 'N': 10,
                                      'combat_style': Melee, 'level': 3, 'num_agent': 1})
  profile_env_step(tasks=tasks)
```

2.0 branch
```
 - env.step({}): 8.587407042999985
 - env.realm.step(): 1.891514784998435
 - env._compute_observations(): 0.5926610509995953
 - obs.to_gym(), ActionTarget: 0.9907544480010984
 - env._compute_rewards(): 3.858267388000968
```

this branch
```
 - env.step({}): 8.907805581999128
 - env.realm.step(): 1.369155314998352
 - env._compute_observations(): 0.5924305849985103
 - obs.to_gym(), ActionTarget: 0.9765266709946445
 - env._compute_rewards(): 4.661505613999907
```

**Comment by kywch** [comment]
Thank you, Mark! I see that the hash patch brought some performance gains in the same `complex_task()` test. Nice job!

```
 - env.step({}): 7.718032001999745
 - env.realm.step(): 1.3791717379999682
 - env._compute_observations(): 0.597285723999903
 - obs.to_gym(), ActionTarget: 0.9713346349999483
 - env._compute_rewards(): 3.569329479999851     VS. 3.858267 (before the hash patch)
```

I agree that the tidying up distance functions, etc, would be better as a separate PR. However, I still have an issue with the `habitable()` function in this PR so cannot move forward without good justification. Thank you.

**Comment by kywch** [comment]
Thanks Mark for making the changes! I agree that the new `habitable()` function is much faster by using `np_random.intergers()` and the cached `habitable_tiles`. However, I think it can be made more readable. For example,

```python
    DIRECTIONS = [ # row delta, col delta, action
        (-1, 0, action.North),
        (1, 0, action.South),
        (0, -1, action.West),
        (0, 1, action.East)] * 2

    def habitable_short(realm_map, ent, np_random):
      r, c = ent.pos
      is_habitable = realm_map.habitable_tiles
      start = np_random.integers(4)
      for i in range(4):
        dr, dc, act = DIRECTIONS[start + i]
        if is_habitable[r + dr, c + dc]:
          return act

      return action.North
```

**Comment by MarkHaoxiang** [comment]
@kywch  Thanks for the suggestion! I have tested and your version seems to have the same performance while being more readable. The corresponding changes have been applied.

**Comment by kywch** [comment]
Thank you for your work, Mark! I'm happy to merge this PR to 2.0

**Review by kywch — APPROVED** [review]

**Pull request merged**